### PR TITLE
ddl: rollingback add index meets panic leads json unmarshal object error

### LINKED ISSUE #23321: ddl: rollingback add index meets panic will lead json unmarshal object error
## Bug Report

Please answer these questions before submitting your issue. Thanks!
https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_check/detail/tidb_ghpr_check/75909/pipeline/

On every occasional way, when I cherry-pick the pr to release-4.0, I meet a very strange test failure with message info like 
```
[2021-03-15T07:13:17.203Z] db_integration_test.go:1308:
[2021-03-15T07:13:17.203Z]     s.tk.MustGetErrCode(sql, errno.ErrDupEntry)
[2021-03-15T07:13:17.203Z] /home/jenkins/agent/workspace/tidb_ghpr_check/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:300:
[2021-03-15T07:13:17.203Z]     tk.c.Assert(int(sqlErr.Code), check.Equals, errCode, check.Commentf("Assertion failed, origin err:\n  %v", sqlErr))
[2021-03-15T07:13:17.203Z] ... obtained int = 1105
[2021-03-15T07:13:17.203Z] ... expected int = 1062
[2021-03-15T07:13:17.203Z] ... Assertion failed, origin err:
[2021-03-15T07:13:17.203Z]   ERROR 1105 (HY000): json: cannot unmarshal object into Go value of type bool
``` 
At the first beginning, I thought it has nothing to do with my PR. After several hours of investigating it, I found it is a quite rare test case triggered by my global failpoint panic in the DDL's drop index logic, which should only be used in the serialTestSuite but I forgot.

### 1. Minimal reproduce step (Required)

**Anyway, thanks to the misused failpoint, we found a very wired DDL problem, caused by two different canceling DDL entrances.**

Here is the error's triggering logic.

1: when add-index meets error in reorg state (which could be [kv:1062]Duplicate entry )

2: since this error could **NOT** be ignored, we convert this add-index to a rollingback job (which will rewrite the `job.args` and `schema-state` for drop-index logic, and set the `job.state` as `JobStateRollingback` )

3: In the next DDL round, the add-index job with `JobStateRollingback` state will walk through the rollingback logic in `onCreateIndex` function, which will step into `onDropIndex`.

4: If there is panic in the drop-index logic, this ddl job will be pulled up and set the `job.state` as canceling (but its `job.args` has already been rewritten for rollingback logic)

5: in the next DDL round, this half-way rollingback job will be recognized as a brand-new canceling job again, which will step into unified canceling entrance `convertJob2RollbackJob` and it will take it for granted that it's args is what it used to be, because it's job.Type is add-index. The decoding error is finally stored as a job error here.

### 2. What did you expect to see? (Required)
A job that has been pulled up from a panic should be handled correctly. Maybe we should unify the canceling entrance.

### 3. What did you see instead (Required)
A very strange decoding error for DDL failure and the remained index hasn't been cleaned either, which is hard to locate and reproduced. 

### 4. What is your TiDB version? (Required)

master, 4.0...

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #23321 

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: ddl/db_test.go [hunk 1]
```diff
@@ -5039,3 +5039,23 @@ func (s *testSerialDBSuite) TestDDLExitWhenCancelMeetPanic(c *C) {
 	c.Assert(job.ErrorCount, Equals, int64(4))
 	c.Assert(job.Error.Error(), Equals, "[ddl:-1]panic in handling DDL logic and error count beyond the limitation 3, cancelled")
 }
+
+// Close issue #23321
+// https://github.com/pingcap/tidb/issues/23321
```
Comment (zimulala) [inline ddl/db_test.go:5044]:
```suggestion
// See https://github.com/pingcap/tidb/issues/23321
```

File: ddl/db_test.go [hunk 2]
```diff
@@ -5039,3 +5039,23 @@ func (s *testSerialDBSuite) TestDDLExitWhenCancelMeetPanic(c *C) {
 	c.Assert(job.ErrorCount, Equals, int64(4))
 	c.Assert(job.Error.Error(), Equals, "[ddl:-1]panic in handling DDL logic and error count beyond the limitation 3, cancelled")
 }
+
+// Close issue #23321
```
Comment (zimulala) [inline ddl/db_test.go:5043]:
```suggestion
// Close issue #23321.
```

File: ddl/db_test.go [hunk 3]
```diff
@@ -5039,3 +5039,23 @@ func (s *testSerialDBSuite) TestDDLExitWhenCancelMeetPanic(c *C) {
...
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists test_add_index_after_add_col")
+	tk.MustExec("create table test_add_index_after_add_col(a int, b int not null default '0');")
+	tk.MustExec("insert into test_add_index_after_add_col values(1, 2),(2,2);")
+	tk.MustExec("alter table test_add_index_after_add_col add column c int not null default '0';")
+
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/mockExceedErrorLimit", `return(true)`), IsNil)
+	defer func() {
+		failpoint.Disable("github.com/pingcap/tidb/ddl/mockExceedErrorLimit")
```
Comment (xiongjiwei) [inline ddl/db_test.go:5056]: the error should be handled, or the error check linter will fail

**Comment by xhebox** [comment]
I did not follow the issue, does the problem exist on the master? Should we file PR merged to the master?

**Comment by xiongjiwei** [comment]
refer https://github.com/pingcap/tidb/issues/23893

**Comment by AilinKid** [comment]
> /cc @AilinKid PTAL. Do we decided how to fix the issue?

I think just to check it in here is adequate for now

**Comment by AilinKid** [comment]
actually, the master can also get the phenomenon above, but the error returned to the statement is substituted by another commit. So we can't see the JSON unmarshal error directly, but if we debug the internal step, we can see the intermediate error is also `JSON unmarshal error`. 

so we can apply the same logic to the master @Howie59

**Comment by Howie59** [comment]
> actually, the master can also get the phenomenon above, but the error returned to the statement is substituted by another commit. So we can't see the JSON unmarshal error directly, but if we debug the internal step, we can see the intermediate error is also `JSON unmarshal error`.
> 
> so we can apply the same logic to the master @Howie59

i submit it in issue #23913

**Review by zimulala — COMMENTED** [review]
/LGTM

**Comment by zimulala** [comment]
PTAL @AilinKid

**Comment by zimulala** [comment]
@Howie59 
Could we merge #23913 first? Then we cherry-pick it to v4.0 and v5.0.

**Review by AilinKid — COMMENTED** [review]
/rest LGTM

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -585,7 +585,14 @@ func chooseLeaseTime(t, max time.Duration) time.Duration {
 // countForPanic records the error count for DDL job.
 func (w *worker) countForPanic(job *model.Job) {
 	// If run DDL job panic, just cancel the DDL jobs.
-	job.State = model.JobStateCancelling
+	if job.State == model.JobStateRollingback {
+		job.State = model.JobStateCancelled
+		msg := fmt.Sprintf("DDL job cancelled by panic in rollingback, error msg: %s", terror.ToSQLError(job.Error).Message)
+		job.Error = terror.GetErrClass(job.Error).Synthesize(terror.ErrCode(job.Error.Code()), msg)
+		logutil.Logger(w.logCtx).Warn(msg)
```
Comment (AilinKid) [inline ddl/ddl_worker.go:592]: sorry, I forget one thing, we can return directly here.

**Review by AilinKid — COMMENTED** [review]
/lgtm

**Comment by AilinKid** [comment]
you can make the check lint happy with
```
if job.State == model.JobStateRollingback {
		job.State = model.JobStateCancelled
		msg := fmt.Sprintf("DDL job cancelled by panic in rollingback, error msg: %s", terror.ToSQLError(job.Error).Message)
		job.Error = terror.GetErrClass(job.Error).Synthesize(terror.ErrCode(job.Error.Code()), msg)
		logutil.Logger(w.logCtx).Warn(msg)
		return
	}
	job.State = model.JobStateCancelling
	job.ErrorCount++
```

**Comment by zimulala** [comment]
@Howie59 
Please address comments.

**Pull request merged**